### PR TITLE
Implement the built in function str_to_date

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3203
issue #5204
issue #1873

## What this PR does / why we need it:
Implement the built-in function str_to_date()

### CODE REVIEW

**Review by aunjgr — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

File: pkg/sql/plan/function/builtin/binary/generalTime.go [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+package binary
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+)
+
+// GeneralTime is the internal struct type for Time.
+type GeneralTime struct {
```
Comment (ouyuanning) [inline pkg/sql/plan/function/builtin/binary/generalTime.go:24]: you should rename generalTime.go to general_time.go

File: pkg/sql/plan/base_binder.go [hunk 2]
```diff
@@ -904,6 +905,24 @@ func bindFuncExprImplByPlanExpr(name string, args []*Expr) (*plan.Expr, error) {
...
+	case "str_to_date", "to_date":
+		if len(args) != 2 {
+			return nil, moerr.NewInvalidArg(name+" function have invalid input args length", len(args))
+		}
+
+		if args[1].Typ.Id == int32(types.T_varchar) || args[1].Typ.Id == int32(types.T_char) {
+			if exprC, ok := args[1].Expr.(*plan.Expr_C); ok {
+				sval := exprC.C.Value.(*plan.Const_Sval)
+				tp, _ := binary.JudgmentToDateReturnType(sval.Sval)
+				args = append(args, makePlan2DateConstNullExpr(tp))
```
Comment (ouyuanning) [inline pkg/sql/plan/base_binder.go:917]: I don't think you use NullExpr is a good choice. Null is null.

May be you can use  Int64Expr， different value for different type.  like:  1=Date 2=Datetime 3=...

File: pkg/sql/plan/make.go [hunk 3]
```diff
@@ -35,6 +35,20 @@ func makePlan2DecimalExprWithType(v string, isBin ...bool) (*plan.Expr, error) {
 	return appendCastBeforeExpr(makePlan2StringConstExprWithType(v, isBin...), typ)
 }
 
+func makePlan2DateConstNullExpr(t types.T) *plan.Expr {
```
Comment (ouyuanning) [inline pkg/sql/plan/make.go:38]: that's not a good choice. you can try another way.

File: pkg/sql/plan/function/builtin/binary/generalTime.go (hunk 1 above)
Comment (qingxinhome) [reply pkg/sql/plan/function/builtin/binary/generalTime.go:24]: OK，no problem!

**Pull request merged**